ksqldb-execution/src/main/java/io/confluent/ksql/execution/codegen/SqlToJavaVisitor.java

```diff
@@ -448,49 +452,108 @@ private String formatExpression(final Expression expression) {
-          }
+          final TypeContext childContext = context.getCopy();
+          final SqlType resolvedArgType = expressionTypeManager.getExpressionSqlType(
+              expression,
+              childContext
+          );
+          argTypesForFirstPass.add(SqlArgument.of(resolvedArgType));
         }
       }
 
-      final KsqlFunction function = udfFactory.getFunction(argumentSchemas);
+      final UdfFactory udfFactory = functionRegistry.getUdfFactory(node.getName());
+      final KsqlScalarFunction function = udfFactory.getFunction(argTypesForFirstPass);
+
+      final List<ParamType> paramTypes = function.parameters();
+      SqlType functionReturnSchema;
+      if (node.hasLambdaFunctionCallArguments()) {
+        final Map<GenericType, SqlType> reservedGenerics = new HashMap<>();
+        final List<SqlArgument> argForReturnType = new ArrayList<>();
+        for (int i = 0; i < arguments.size(); i++) {
+          final Expression expression = arguments.get(i);
+          final ParamType parameter = paramTypes.get(i);
+          if (expression instanceof LambdaFunctionCall) {
+            final TypeContext childContext = context.getCopy();
+
+            // the function returned from the UDF factory should have lambda
+            // at this index in the function arguments if there's a
+            // lambda node at this index in the function node argument list
+            if (!(parameter instanceof LambdaType)) {
+              throw new KsqlException("Error while processing lambda function. "
+                  + "This is most likely an internal error and a "
+                  + "Github issue should be filed for debugging.");
+            }
+
+            final LambdaType lambdaParameter = (LambdaType) parameter;
+            for (ParamType inputParam : lambdaParameter.inputTypes()) {
+              if (inputParam instanceof GenericType) {
+                final GenericType genericParam = (GenericType) inputParam;
+                if (!reservedGenerics.containsKey(genericParam)) {
+                  throw new KsqlException(
+                      String.format(
+                          "Could not verify type for generic %s.",
+                          genericParam.toString()));
+                }
+                childContext.addLambdaInputType(reservedGenerics.get(genericParam));
+              } else {
+                childContext.addLambdaInputType(
+                    SchemaConverters.functionToSqlConverter().toSqlType(inputParam));
+              }
+            }
+            final List<SqlType> sqlTypesForLambdaArgMapping =
+                new ArrayList<>(childContext.getLambdaInputTypes());
+            final SqlType resolvedArgType = expressionTypeManager.getExpressionSqlType(
+                expression,
+                childContext.getCopy()
+            );
+            final SqlArgument lambdaArgument = SqlArgument.of(
+                SqlLambda.of(sqlTypesForLambdaArgMapping, resolvedArgType));
```

Comment:
Why did you do `sqlTypesForLambdaArgMapping` instead of doing `childContext.getLambdaInputTypes()`? Since we passs in a copy to `ExpressionTypeManager` won't the input type list be unchanged when we create the sqlLambda?

Reply:
This was from an earlier iteration where the input type list was getting wiped while handling the child, so I added this in. Removed now.